crates/store/re_datafusion/src/table_entry_provider.rs

```diff
@@ -154,4 +176,248 @@ impl GrpcStreamToTable for TableEntryTableProvider {
+    grpc_error: Option<tonic::Status>,
+    insert_op: TableInsertMode,
+}
+
+struct GrpcStreamSender {
+    sender: tokio::sync::mpsc::UnboundedSender<WriteTableRequest>,
+}
+
+impl RecordBatchStream for RecordBatchGrpcOutputStream {
+    fn schema(&self) -> SchemaRef {
+        Arc::new(Schema::empty())
+    }
+}
+
+impl RecordBatchGrpcOutputStream {
+    pub fn new(
+        input_stream: SendableRecordBatchStream,
+        client: ConnectionClient,
+        runtime: &Handle,
+        table_id: EntryId,
+        insert_op: TableInsertMode,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        // Create an oneshot channel for reporting when the thread is complete
+        let (thread_status_tx, thread_status_rx) = tokio::sync::oneshot::channel();
+
+        runtime.spawn(async move {
+            let shutdown_response = async {
+                let stream = tokio_stream::wrappers::UnboundedReceiverStream::new(rx)
+                    .into_streaming_request()
+                    .with_entry_id(table_id)?;
+                let mut client = client;
+
+                client.write_table(stream).await
+            }
+            .await;
+
+            #[expect(clippy::let_underscore_must_use)]
+            let _ = thread_status_tx.send(shutdown_response);
+        });
+
+        Self {
+            input_stream,
+            grpc_sender: Some(GrpcStreamSender { sender: tx }),
+            thread_status: thread_status_rx,
+            complete: false,
+            grpc_error: None,
+            insert_op,
+        }
+    }
+}
+
+impl Stream for RecordBatchGrpcOutputStream {
+    type Item = Result<RecordBatch, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check for gRPC errors first (only if we haven't already stored one)
+        if self.grpc_error.is_none() {
+            match Pin::new(&mut self.thread_status).poll(cx) {
+                Poll::Ready(Ok(Err(status))) => {
+                    // Store the error for potential future use
+                    self.grpc_error = Some(status.clone());
+                    // Return the error immediately
+                    return Poll::Ready(Some(Err(DataFusionError::External(Box::new(status)))));
+                }
+                Poll::Ready(Ok(Ok(())) | Err(_)) => {
+                    self.complete = true;
+                }
+                Poll::Pending => {}
+            }
+        }
+
+        match Pin::new(&mut self.input_stream).poll_next(cx) {
+            Poll::Ready(Some(Ok(batch))) => {
+                // Send to gRPC if we have a sender
+                if let Some(ref grpc_sender) = self.grpc_sender {
+                    let dataframe_part = batch.into();
+                    let request = WriteTableRequest {
+                        dataframe_part: Some(dataframe_part),
+                        insert_mode: self.insert_op.into(),
+                    };
```

Comment:
with my proposed `ConnectionClient` change, we'd be moving `RecordBatch`es here, which feels more logical than obscure grpc types.